processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -218,18 +219,10 @@ private void sortPropertyMappingsByDependencies() {
             else {
                 Collections.sort(
                     propertyMappings, new Comparator<PropertyMapping>() {
-
                         @Override
                         public int compare(PropertyMapping o1, PropertyMapping o2) {
-                            if ( graphAnalyzer.getAllDescendants( o1.getName() ).contains( o2.getName() ) ) {
-                                return 1;
-                            }
-                            else if ( graphAnalyzer.getAllDescendants( o2.getName() ).contains( o1.getName() ) ) {
-                                return -1;
-                            }
-                            else {
-                                return 0;
-                            }
```

Comment:
This old implementation of the `Comparator` violates the API spec in this paragraph, which led to the bug:

> Finally, the implementor must ensure that `compare(x, y)==0` implies that `sgn(compare(x, z))==sgn(compare(y, z))` for all `z`.